### PR TITLE
Create simplified api request helper

### PR DESCRIPTION
**Pull request by phil303**
The motivation for this patch initially was to remove some of the abstraction from our api layer to gain me access to the http status codes. It then bloomed into a complete refactor of APIRequestUtils before I decided to reel the scope back in. What it is now is a single helper function that can do the work of `runForm`, `runQuery`, and `runJson` at a much lower abstraction level that deprecates those guys.

By design it does very little "handling" of the response for you but for consistency and simplicity does still hand you off an `ApiResponse` object. This object now has references to the initial request and its response so you can access any lower level things like the status codes should the need arise.

Ultimately, I'd like to give all of our api wrappers this treatment. The attempt to do that in this patch was just too much change in one go so I'll have to come back to it.

:eyeglasses: @schwers

### CODE REVIEW

File: apiBase/apiRequest.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  if (_method === 'post') {
+    request.send(body);
+  }
+
+  return new Promise((resolve, reject) => {
+    request.end((err, res) => {
+      if (!err) {
+        resolve(new APIResponse(res));
+      } else {
+        reject(err);
```
Comment (phil303) [inline apiBase/apiRequest.js]: per convo with Ryan, we should reject `ResponseError` here to normalize superagent's error responses.

File: apiBase/apiRequest.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import superagent from 'superagent';
+
+import APIResponse from './APIResponse';
+
+
+export default (apiOptions, method, path, query={}, body={}, type=null) => {
```
Comment (phil303) [inline apiBase/apiRequest.js]: consider making query, body, and type into one optional argument

**Comment by phil303** [comment]
:haircut:

File: apiBase/apiRequest.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import APIResponse from './APIResponse';
+import ResponseError from './errors/ResponseError';
+
+
+/* A thin helper function around our apis.
+ *
+ * @param {Object} apiOptions - fields required to hit our api
+ * @param {String} method   - the http method upper or lowercased, e.g. 'GET'
+ * @param {String} path     - the endpoint path
+ * @param {String} options  - use to set query params, a body, or request type
```
Comment (nramadas) [inline apiBase/apiRequest.js]: `* @param {Object} options`

Comment (phil303) [inline apiBase/apiRequest.js]: fixing, thanks!

File: apiBase/apiRequest.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ */
+export default (apiOptions, method, path, options={}) => {
+  const { query={}, body={}, type=null } = options;
+  const { origin, appName, env, token, defaultHeaders={} } = apiOptions;
+
+  const _method = method.toLowerCase();
+  const _headers = token ?
```
Comment (nramadas) [inline apiBase/apiRequest.js]: for consistency, should we format this multiline ternary like we do the others?
```es6
const _header = token
  ? { ...defaultHeaders, Authorization: `Bearer ${token}` }
  : defaultHeaders;
```

Comment (phil303) [inline apiBase/apiRequest.js]: good idea. updating.

File: apis/PostsEndpoint.js [hunk 5]
```diff
@@ -60,41 +61,41 @@ const formatPostData = (data)=> {
...
 export default {
-  get(apiOptions, query) {
-    const path = getPath(query);
-    const apiQuery = formatQuery({...query}, 'get');
+  get(apiOptions, _query) {
+    const path = getPath(_query);
+    const query = formatQuery({..._query}, 'get');
 
-    return runQuery(apiOptions, 'get', path, apiQuery, query, handleGet);
+    return apiRequest(apiOptions, 'GET', path, { query }).then(handleGet);
```
Comment (nramadas) [inline apis/PostsEndpoint.js]: out of curiosity, why do you uppercase the methods?

File: apis/PostsEndpoint.js [hunk 6]
```diff
@@ -60,41 +61,41 @@ const formatPostData = (data)=> {
...
+  return apiResponse;
 };
 
 export default {
-  get(apiOptions, query) {
-    const path = getPath(query);
-    const apiQuery = formatQuery({...query}, 'get');
+  get(apiOptions, _query) {
+    const path = getPath(_query);
+    const query = formatQuery({..._query}, 'get');
```
Comment (nramadas) [inline apis/PostsEndpoint.js]: the linter will probably catch this, but spaces to pad the `{}`

File: apis/PostsEndpoint.js (hunk 5 above)
Comment (phil303) [inline apis/PostsEndpoint.js]: couple reasons - none i feel inherently strongly about.
1) Stands out. They could probably be constants but that makes the function slightly more unwieldy.
2) That's the way I'm used to seeing them, in the http spec, when I write stuff

**Comment by nramadas** [comment]
:nail_care:

**Comment by nramadas** [comment]
From a high level view, I think this is a great improvement. Just a few small nits, other than that it looks good.

File: apis/PostsEndpoint.js (hunk 6 above)
Comment (phil303) [inline apis/PostsEndpoint.js]: ooo thank you

**Comment by phil303** [comment]
thanks! updated :haircut:

**Comment by nramadas** [comment]
:fish:

**Pull request merged**